src/consensus/tx_verify.cpp

```diff
@@ -235,35 +234,16 @@ bool CheckZerocoinSpend(const CTransaction& tx, CValidationState& state)
     return fValidated;
 }
 
-// Create a lru cache to hold the currently validated pubcoins with a max size of 5000
-LRUCacheTemplate<std::string,bool> cacheValidatedPubcoin(5000);
-CCriticalSection cs_pubcoinCache;
 bool CheckZerocoinMint(const CTxOut& txout, CBigNum& bnValue, CValidationState& state, bool fSkipZerocoinMintIsPrime)
 {
     libzerocoin::PublicCoin pubCoin(Params().Zerocoin_Params());
     if (!TxOutToPublicCoin(txout, pubCoin))
         return state.DoS(100, error("CheckZerocoinMint(): TxOutToPublicCoin() failed"));
 
-    bnValue = pubCoin.getValue();
-    uint256 hashPubcoin = GetPubCoinHash(bnValue);
-
-    //CheckZerocoinMint can be done from different threads at the same time. Lock static containers. If try_lock fails,
-    //don't wait, just do full validation.
-    TRY_LOCK(cs_pubcoinCache, fLocked);
-    bool value;
-    if (!fSkipZerocoinMintIsPrime) {
-        bool fRunValidation = !fLocked || (fLocked && !cacheValidatedPubcoin.get(hashPubcoin.GetHex(), value));
-        if (fRunValidation) {
-            if (!pubCoin.validate())
-                return state.DoS(100, error("CheckZerocoinMint() : PubCoin does not validate"));
-            if (fLocked) {
-                cacheValidatedPubcoin.set(hashPubcoin.GetHex(), true);
-            }
-        }
-
-
-    }
+    if (!pubCoin.validate())
+        return state.DoS(100, error("CheckZerocoinMint() : PubCoin does not validate"));
 
+    bnValue = pubCoin.getValue();
```

Comment:
This does change a behavior.... before: regardless of if pubCoin validated or not, bnValue is set.  Now, if the validation fails, bnValue remains unchanged.  But we need to make absolutely sure that there isn't any code that uses bnPubCoin this filters back up unchanged.

We should document the routine to explain the conditions bnValue is set, and the conditions bnValue is not set.

The change appears to be fine (initial inspection seems that CheckZerocoinMint() is only called from CheckStandardOutput, which is only called from CheckTransaction() which returns false in CheckStandardOutput fails....   so it -should- be okay.  But that due diligence is needed (both in the details of this PR, and the addition of a little documentation on this routine to explain the conditions that change the arguments.

Reply:
I agree.. I reverted the function back to the current logic and propose we go ahead with this.